src/laser/MultiLaser.cpp

```diff
@@ -534,50 +589,216 @@ MultiLaser::ShiftLaserSlices ()
+        const amrex::Real poff_field_y = GetPosOffset(1, geom_field_lev0, geom_field_lev0.Domain());
+
+        const amrex::Real dx_laser = m_laser_geom_3D.CellSize(0);
+        const amrex::Real dy_laser = m_laser_geom_3D.CellSize(1);
+        const amrex::Real dx_laser_inv = m_laser_geom_3D.InvCellSize(0);
+        const amrex::Real dy_laser_inv = m_laser_geom_3D.InvCellSize(1);
+        const amrex::Real dx_field = geom_field_lev0.CellSize(0);
+        const amrex::Real dy_field = geom_field_lev0.CellSize(1);
+        const amrex::Real dx_field_inv = geom_field_lev0.InvCellSize(0);
+        const amrex::Real dy_field_inv = geom_field_lev0.InvCellSize(1);
+
+        amrex::Box field_box = fields.getSlices(0)[mfi].box();
+        field_box.grow(-2*Fields::m_slices_nguards);
+
+        const amrex::Real pos_x_lo = field_box.smallEnd(0) * dx_field + poff_field_x;
+        const amrex::Real pos_x_hi = field_box.bigEnd(0) * dx_field + poff_field_x;
+        const amrex::Real pos_y_lo = field_box.smallEnd(1) * dy_field + poff_field_y;
+        const amrex::Real pos_y_hi = field_box.bigEnd(1) * dy_field + poff_field_y;
+
+        // the indexes of the laser box where the fields box ends
+        const int x_lo = amrex::Math::ceil((pos_x_lo - poff_laser_x) * dx_laser_inv);
+        const int x_hi = amrex::Math::floor((pos_x_hi - poff_laser_x) * dx_laser_inv);
+        const int y_lo = amrex::Math::ceil((pos_y_lo - poff_laser_y) * dy_laser_inv);
+        const int y_hi = amrex::Math::floor((pos_y_hi - poff_laser_y) * dy_laser_inv);
+
+        amrex::ParallelFor(
+            amrex::TypeList<amrex::CompileTimeOptions<0, 1, 2, 3>>{},
+            {m_interp_order},
+            mfi.growntilebox(),
+            [=] AMREX_GPU_DEVICE(int i, int j, int, auto interp_order) noexcept {
+                const amrex::Real x = i * dx_laser + poff_laser_x;
+                const amrex::Real y = j * dy_laser + poff_laser_y;
+
+                const amrex::Real xmid = (x - poff_field_x) * dx_field_inv;
+                const amrex::Real ymid = (y - poff_field_y) * dy_field_inv;
+
+                amrex::Real chi = 0;
+
+                if (x_lo <= i && i <= x_hi && y_lo <= j && j <= y_hi) {
+                    // interpolate chi from fields to laser
+                    for (int iy=0; iy<=interp_order; ++iy) {
+                        for (int ix=0; ix<=interp_order; ++ix) {
+                            auto [shape_x, cell_x] =
+                                compute_single_shape_factor<false, interp_order>(xmid, ix);
+                            auto [shape_y, cell_y] =
+                                compute_single_shape_factor<false, interp_order>(ymid, iy);
+
+                            chi += shape_x*shape_y*field_arr_chi(cell_x, cell_y);
+                        }
+                    }
+                } else {
+                    // get initial chi outside the fields box
+                    chi = laser_arr(i, j, WhichLaserSlice::chi_initial);
```

Comment:
This is labeled initial, but is still recomputed at every time step to account for the z variation of the density profile, right?